app/services/title_builder.rb

```diff
@@ -4,38 +4,130 @@ class TitleBuilder
+      return title_from_structured_values(structured_value.structuredValue, non_sorting_char_count) if structured_value.structuredValue
+
+      value = structured_value.value&.strip
+      next unless value
+
+      # additional types:  name, uniform ...
+      case structured_value.type&.downcase
+      when 'nonsorting characters'
+        non_sort_value = value&.size == non_sorting_char_count ? value : "#{value} "
+        structured_title = if structured_title.present?
+                             "#{structured_title}#{non_sort_value}"
+                           else
+                             non_sort_value
+                           end
+      when 'part name', 'part number'
+        if title_from_part.blank?
+          title_from_part = title_from_structured_part(structured_values)
+          structured_title = if structured_title.present?
+                               "#{structured_title.sub(/[ .,]*$/, '')}. #{title_from_part}. "
+                             else
+                               "#{title_from_part}. "
+                             end
+        end
+      when 'main title', 'title'
+        structured_title = "#{structured_title}#{value}"
+      when 'subtitle'
+        # subtitle is preceded by space colon space, unless it is at the beginning of the title string
+        structured_title = if structured_title.present?
+                             "#{structured_title.sub(/[. :]+$/, '')} : #{value.sub(/^:/, '').strip}"
+                           else
+                             structured_title = value.sub(/^:/, '').strip
+                           end
+      end
+    end
+    structured_title
+  end
+  # rubocop:enable Metrics/BlockLength
+  # rubocop:enable Metrics/CyclomaticComplexity
+  # rubocop:enable Metrics/MethodLength
+  # rubocop:enable Metrics/PerceivedComplexity
+  private_class_method :title_from_structured_values
+
+  def self.remove_trailing_punctuation(title)
+    title.sub(%r{[ .,;:/\\]+$}, '')
+  end
+  private_class_method :remove_trailing_punctuation
+
+  # @param [Array<Cocina::Models::Title>] titles
+  # @return [Cocina::Models::Title, nil] title that has status=primary
+  def self.primary_title(titles)
+    primary_title = titles.find do |title|
+      title.status == 'primary'
+    end
+    return primary_title if primary_title.present?
+
+    # NOTE: structuredValues would only have status primary assigned as a sibling, not as an attribute
+
+    titles.find do |title|
+      title.parallelValue&.find do |parallel_title|
+        parallel_title.status == 'primary'
+      end
+    end
+  end
+  private_class_method :primary_title
+
+  # @param [Array<Cocina::Models::Title>] titles
+  # @return [Cocina::Models::Title, nil] first title that has no type attribute
+  def self.first_untyped_title(titles)
+    titles.find do |title|
+      if title.parallelValue.present?
+        title.parallelValue&.find { |parallel_value| parallel_value.type.nil? }
```

Comment:
you just checked that parallel value was present, so why use `parallelValue&.find`?

Reply:
good point; removing.